Release/v0.2.10

thomashebrard opened the pull request:
### Highlights

**Python Support Expansion** - We're no longer tied to Python 3.11! Now supporting Python 3.11, 3.12, and 3.13 with full CI coverage across all versions.

**Major Model Additions** - Claude 4 (Opus & Sonnet), Grok-3, and GPT-4 image generation are now in the house.

### Pipeline Base Library update
- **New pipe** - `extract_page_contents_and_views_from_pdf` transfered from cookbook to base library (congrats on the promotion!). This pipe extracts text, linked images, **AND** page_view images (rendered pages) - it's very useful if you want to use Vision in follow-up pipes

### Added

- **Template preprocessor** - New `@?` token prefix for optional variable insertion - if a variable doesn't exist, we gracefully skip it instead of throwing exceptions
- **Claude 4 support** - Both Opus and Sonnet variants, available through Anthropic SDK (direct & Bedrock) plus Bedrock SDK. Includes specific max_tokens limit reduction to prevent timeout/streaming issues (temporary workaround)
- **Grok-3 family support** - Full support via OpenAI SDK for X.AI's latest models  
- **GPT-4 image generation** - New `gpt-image-1` model through OpenAI SDK, available via PipeImgGen. Currently saves local files (addressing in next release)
- **Gemini update** - Added latest `gemini-2.5-pro` to the lineup
- **Image generation enhancements** - Better quality controls, improved background handling options, auto-adapts to different models: Flux, SDXL and now gpt-image-1

### Refactored

- Moved subpackage `plugin` to the same level as `cogt` within **pipelex** for better visibility
- Major cleanup in the unit tests, hierarchy significantly flattened
- Strengthened error handling throughout inference flows and template preprocessing
- Added `make test-quiet` (shorthand `tq`) to Makefile to run tests without capturing outputs (i.e. without pytest `-s` option)
- Stopped using Fixtures for `pipe_router` and `content_generator`: we're now always getting the singleton from `pipelex.hub`


### Fixed

- **Perplexity integration** - Fixed breaking changes from recent updates

### Dependencies

- Added **pytest-xdist** to run unit tests in parallel on multiple CPUs. Not yet integrated into the Makefile, so run it manually with `pytest -n auto` (without inference) or `pytest -n auto -m "inference"` (inference only). 
- Swapped pytest-pretty for pytest-sugar - because readable test names > pretty tables
- Updated instructor to v1.8.3
- All dependencies tested against Python 3.11, 3.12, and 3.13

### Tests

- TestTemplatePreprocessor
- TestImggByOpenAIGpt
- TestImageGeneration
- TestPipeImgg

### 🔄 Type of Change
- [x] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [x] 📚 Documentation update
- [x] 🧹 Code refactor
- [x] ⚡ Performance improvement
- [x] ✅ Test update

antoinepipelex reviewed (approved):
ok

thomashebrard committed:
fix/guard-branch

thomashebrard committed:
fix/guard-branch (#26)

thomashebrard committed:
bump-version to 0.2.10

thomashebrard committed:
Merge branch 'release/v0.2.10' into dev

thomashebrard committed:
Merge pull request #27 from Pipelex/dev

dev to release/v0.2.10

thomashebrard committed:
fix/python3>=11 (#24)

* fix/python3>=11

* f

* fix makefile

* gha

* fix

* test

* e

* a

* ok

* t

* a

* lint

thomashebrard committed:
Fix/makefile python version (#30)

* fix/tests-without-google cred

* fix/Makefile-python-version

* fix Makefile

lchoquel committed:
Add inference models & template optional insertion (#32)

### Features
- Template preprocessor new token prefix `@?` for **optional insertion**: if the variable doesn't exist in the context, nothing is inserted (instead of raising an exception).

### Models
- Added **Claude 4 Opus and Sonnet**, usable through the Anthropic SDK at Anthropic and Bedrock and usable with the Bedrock SDK. Includes a specific setting to reduce the max_tokens limit to avoid a timeout/streaming issue yet to be addressed later.
- Added support for the **Grok 3** family through the OpenAI SDK.
- Added support for **GPT4o image generation** through the OpenAI SDK, with the gpt-image-1 model. Available through PipeImgGen as an alternative to the Forest Lab (FAL) supported models. Currently saves local files, to be addressed.
- Added latest gemini-2.5-pro

### Base Library
- **Added Pipe:** extract_page_contents_and_views_from_pdf which not only grabs the text and linked images from PDF but also includes `page_view` images, i.e. pages rendered as images (very useful for using Vision down the line). This one used to be defined in the cookbook samples so, congrats on the promotion.

### Refactor
- Improved **error handling** in several places regarding inference
- Moved cogt tests into the tests/pipelex so they can share fixtures, currently used for imgg_handle (image generation handle)

### Fixes
- Fixed integration of **Perplexity**, broken by recent changes

### Dependencies
- Updated **instructor** to 1.8.3
- Dev: replaced pytest-pretty by **pytest-sugar** because pytest-pretty truncates test-case names just to make it fit into a pretty rich table. Function before fashion.

### 🔄 Type of Change
- [x] 🐛 Bug fix
- [X] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation update
- [X] 🧹 Code refactor
- [ ] ⚡ Performance improvement
- [X] ✅ Test update

### 🧪 Tests
- TestTemplatePreprocessor
- TestImggByOpenAIGpt
- TestImageGeneration
- TestPipeImgg

thomashebrard committed:
Merge branch 'release/v0.2.10' into dev

thomashebrard committed:
dev to release/v0.2.10 (#31)

* fix/python3>=11 (#24)

* fix/python3>=11

* f

* fix makefile

* gha

* fix

* test

* e

* a

* ok

* t

* a

* lint

* Fix/makefile python version (#30)

* fix/tests-without-google cred

* fix/Makefile-python-version

* fix Makefile

* Add inference models & template optional insertion (#32)

### Features
- Template preprocessor new token prefix `@?` for **optional insertion**: if the variable doesn't exist in the context, nothing is inserted (instead of raising an exception).

### Models
- Added **Claude 4 Opus and Sonnet**, usable through the Anthropic SDK at Anthropic and Bedrock and usable with the Bedrock SDK. Includes a specific setting to reduce the max_tokens limit to avoid a timeout/streaming issue yet to be addressed later.
- Added support for the **Grok 3** family through the OpenAI SDK.
- Added support for **GPT4o image generation** through the OpenAI SDK, with the gpt-image-1 model. Available through PipeImgGen as an alternative to the Forest Lab (FAL) supported models. Currently saves local files, to be addressed.
- Added latest gemini-2.5-pro

### Base Library
- **Added Pipe:** extract_page_contents_and_views_from_pdf which not only grabs the text and linked images from PDF but also includes `page_view` images, i.e. pages rendered as images (very useful for using Vision down the line). This one used to be defined in the cookbook samples so, congrats on the promotion.

### Refactor
- Improved **error handling** in several places regarding inference
- Moved cogt tests into the tests/pipelex so they can share fixtures, currently used for imgg_handle (image generation handle)

### Fixes
- Fixed integration of **Perplexity**, broken by recent changes

### Dependencies
- Updated **instructor** to 1.8.3
- Dev: replaced pytest-pretty by **pytest-sugar** because pytest-pretty truncates test-case names just to make it fit into a pretty rich table. Function before fashion.

### 🔄 Type of Change
- [x] 🐛 Bug fix
- [X] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation update
- [X] 🧹 Code refactor
- [ ] ⚡ Performance improvement
- [X] ✅ Test update

### 🧪 Tests
- TestTemplatePreprocessor
- TestImggByOpenAIGpt
- TestImageGeneration
- TestPipeImgg

---------

Co-authored-by: Louis Choquel <lchoquel@users.noreply.github.com>

lchoquel committed:
Changelog

thomashebrard committed:
fix/config-project-name (#33)

thomashebrard committed:
Merge branch 'release/v0.2.10' into dev

thomashebrard committed:
fix/config-project-name (#33) (#34)

lchoquel committed:
Refactor/move plugin package up (#35)

- Moved “plugin” to the same level as “cogt” within **pipelex** for better visibility
- Major cleanup in the unit tests: hierarchy largely flattened
- Added **pytest-xdist** to run unit tests in parallel on multiple CPUs. Not yet integrated into the Makefile, so run it manually with `pytest -n auto` or `pytest -n auto -m "llm"`
- Added `make test-quiet` (shorthand `tq`) to Makefile to run tests without capturing outputs (i.e. without pytest `-s` option)
- Stopped using Fixtures for `pipe_router` and `content_generator`: we're now always getting the singleton from `pipelex.hub`

lchoquel committed:
Merge branch 'dev' into release/v0.2.10

lchoquel committed:
Changelog completed

lchoquel committed:
Completed Changelog

antoinepipelex commented on pipelex/pipelex.toml:
```diff
@@ -135,9 +139,11 @@ is_sync_mode = false
 [cogt.imgg_config.imgg_param_defaults]
 aspect_ratio = "square"  # "square", "landscape_4_3", "landscape_16_9", "landscape_21_9", "portrait_3_4", "portrait_9_16", "portrait_9_21"
-nb_steps = 1  # 28 for Flux, one of [1,2,4,8] for SDXL Lightning
+# nb_steps = 1  # 28 for Flux, one of [1,2,4,8] for SDXL Lightning
```
Why is this commented ?

antoinepipelex commented on pipelex/plugin/fal/fal_factory.py:
```diff
@@ -23,7 +15,27 @@ def make_fal_application(imgg_engine: ImggEngine) -> str:
+    def make_nb_steps_from_quality_for_flux_pro(cls, quality: Quality) -> int:
+        match quality:
+            case Quality.LOW:
+                return 14
```
I think these should be in the config

antoinepipelex commented on pipelex/plugin/fal/fal_factory.py:
```diff
@@ -23,7 +15,27 @@ def make_fal_application(imgg_engine: ImggEngine) -> str:
+    def make_nb_steps_from_quality_for_sdxl_lightning(cls, quality: Quality) -> int:
+        match quality:
+            case Quality.LOW:
+                return 2
```
Same remark

antoinepipelex commented on pipelex/plugin/openai/openai_imgg_worker.py:
```diff
@@ -0,0 +1,92 @@
+        moderation = OpenAIImggFactory.moderation_for_gpt_image_1(is_moderated=imgg_job.job_params.is_moderated)
+        background = OpenAIImggFactory.background_for_gpt_image_1(background=imgg_job.job_params.background)
+        quality = OpenAIImggFactory.quality_for_gpt_image_1(quality=imgg_job.job_params.quality or Quality.LOW)
+        output_compression = 100
```
Should be in the config I think

antoinepipelex reviewed (dismissed):
Epic PR! 
Some non blocking remarks

thomashebrard committed:
Dev to releave/v0.2.10 (#39)



### 📝 Description

- Rename plugins package and move its confg
- Use constants and defaults via configs instead of code magic numbers
* Feature/python>=3.10 (#37)

---------

Co-authored-by: Louis Choquel <lchoquel@users.noreply.github.com>

lchoquel committed:
Updated changelog to show python 3.10

antoinepipelex reviewed (approved):
ok

lchoquel commented:
Epic release
